### PR TITLE
Feature/prelaunch proto refactoring 2016 03 04

### PR DESCRIPTION
**Pull request by cjrd**
This PR includes a fairly large number of updates to the Payment namespace:
* Made all response associated with a request to a payment processor include information about the request, e.g. the Refunded response includes information about the amount refunded, when it was refunded, etc. Most of these responses are fairly redundent right now (we should fix this in the future)
* All responses include a `livemode` field that indicate whether the request was actually handled by the live payment processor
* Extracted redundent enum `PaymentStatus` outside of individual messages
* Switched `payid` in `PreAuthorization` responses to be `hwid` to reflect that these ids are associated with the hardware
* Removed deprecated `FinishedVoid` response
* Updated RPDTS args to be consistent with Hardware.RPD args
* Added `ReceivePaymentDataThenPreAuthorize` request and associated messages
* Large amount of copy rewriting, consistency checks, and removal of deprecated references.

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/DotDashPay/ddp.api.common/pull/19" target="_blank">ReviewNinja

File: protobuf/impls/payment.proto [hunk 1]
```diff
@@ -9,84 +9,203 @@ option objc_class_prefix= "DDP";
...
 
-  Response when an authorization void request is returned from the
+  Response when a void preauthorization response is returned from the
   payment processor.
 */
 message VoidedPreAuthorize {
-}
+  /* an enum with values: SUCCESS = 0, FAIL = 1, UNKNOWN = 2
+     indicating the status of the void request */
+  required PaymentStatus status = 2;
```
Comment (sarietta) [inline protobuf/impls/payment.proto]: Should this reference the PaymentStatus top-level enum instead?

File: protobuf/impls/payment.proto [hunk 2]
```diff
@@ -95,53 +214,113 @@ message StartedSettling {
...
+  optional bool livemode = 15 [default=false]
 }
 
 /*
   @define-response(Settled)
 
-  Response when the settle payment request returns from the payment
-  processor.
+  Response when the @request(payment.Settle) payment request returns
+  from the payment processor.
```
Comment (sarietta) [inline protobuf/impls/payment.proto]: Can you add a note about the fact that the amount in the Settled response is expressed in cents (no dollars)? This seems like something that might confuse people if we're not explicit.

**Comment by sarietta** [comment]
A couple of extra steps need to be performed for this PR to have all of the necessary files needed to actually update the API:

1. Run `spec/check-specification-consistency.js` (no arguments are needed if run in the `spec` directory)
2. Fix any issues raised by (1)
3. Run `spec/generate-protos.js --output_dir=../protobuf` from the `spec` directory, This will likely complain if you haven't adjust the API version numbers in `api-spec.json5`.
4. Commit the new changes to this PR and consider checking in `protobuf/versions/OLD_VERSION` where OLD_VERSION is whatever directory was created for you as a backup. Adding these older version files in is not strictly necessary.

**Comment by cjrd** [comment]
Okay, followed your instructiones and committed results. Let me know if I can do anything else =-)

**Comment by sarietta** [comment]
Here are the compiler failures when I attempt to compile the new protos:

```bash
Protoc DDP Plugin: /usr/local/bin/ddprpc_objc_plugin
Running protoc -I./ddp.api -I/usr/local/include --objc_out=./build .//ddp.api/dotdashpay/api/common/protobuf/api_common.proto
Running protoc -I./ddp.api -I/usr/local/include --objc_out=./build .//ddp.api/dotdashpay/api/common/protobuf/hardware.proto
Running protoc -I./ddp.api -I/usr/local/include --objc_out=./build .//ddp.api/dotdashpay/api/common/protobuf/payment.proto
dotdashpay/api/common/protobuf/payment.proto:89:1: Expected ";".
dotdashpay/api/common/protobuf/payment.proto:153:3: Expected ";".
dotdashpay/api/common/protobuf/payment.proto:212:3: Expected ";".
dotdashpay/api/common/protobuf/payment.proto:272:3: Expected ";".
dotdashpay/api/common/protobuf/payment.proto:297:1: Expected ";".
dotdashpay/api/common/protobuf/payment.proto:361:3: Expected ";".
Running protoc -I./ddp.api -I/usr/local/include --objc_out=./build .//ddp.api/dotdashpay/api/common/protobuf/information.proto
Running protoc -I./ddp.api -I/usr/local/include --objc_out=./build .//ddp.api/dotdashpay/api/common/protobuf/global.proto
Running protoc -I./ddp.api -I/usr/local/include --objc_out=./build .//ddp.api/dotdashpay/api/common/protobuf/datalink.proto
Running protoc -I./ddp.api -I/usr/local/include --plugin=protoc-gen-ddprpc=/usr/local/bin/ddprpc_objc_plugin --ddprpc_out=./ddp.api .//ddp.api/dotdashpay/api/common/protobuf/services.proto
dotdashpay/api/common/protobuf/payment.proto:89:1: Expected ";".
dotdashpay/api/common/protobuf/payment.proto:153:3: Expected ";".
dotdashpay/api/common/protobuf/payment.proto:212:3: Expected ";".
dotdashpay/api/common/protobuf/payment.proto:272:3: Expected ";".
dotdashpay/api/common/protobuf/payment.proto:297:1: Expected ";".
dotdashpay/api/common/protobuf/payment.proto:361:3: Expected ";".
dotdashpay/api/common/protobuf/services.proto: Import "dotdashpay/api/common/protobuf/payment.proto" was not found or had errors.
dotdashpay/api/common/protobuf/services.proto:19:20: "dotdashpay.api.payment.PreAuthorizeArgs" is not defined.
dotdashpay/api/common/protobuf/services.proto:24:14: "dotdashpay.api.payment.SettleArgs" is not defined.
dotdashpay/api/common/protobuf/services.proto:29:24: "dotdashpay.api.payment.VoidPreAuthorizeArgs" is not defined.
dotdashpay/api/common/protobuf/services.proto:33:14: "dotdashpay.api.payment.RefundArgs" is not defined.
dotdashpay/api/common/protobuf/services.proto:37:26: "dotdashpay.api.payment.ConfigureProcessorArgs" is not defined.
dotdashpay/api/common/protobuf/services.proto:41:36: "dotdashpay.api.payment.ReceivePaymentDataThenSettleArgs" is not defined.
dotdashpay/api/common/protobuf/services.proto:48:42: "dotdashpay.api.payment.ReceivePaymentDataThenPreAuthorizeArgs" is not defined.
```

**Pull request merged**